libs/common/RespReadUtils.cs

```diff
@@ -4,156 +4,372 @@
+            if (!TryReadUlong(ref ptr, end, out var number, out var digitsRead))
+            {
                 return false;
+            }
+
+            // Check for overflows and convert digits to int, if possible
+            if (negative)
+            {
+                if (number > ((ulong)int.MaxValue) + 1)
+                {
+                    RespParsingException.ThrowIntegerOverflow(ptr - digitsRead, (int)digitsRead);
+                }
+
+                value = (int)(0 - (long)number);
+            }
+            else
+            {
+                if (number > int.MaxValue)
+                {
+                    RespParsingException.ThrowIntegerOverflow(ptr - digitsRead, (int)digitsRead);
+                }
+                value = (int)number;
+            }
+
+            bytesRead += digitsRead;
 
-            Debug.Assert(*(ptr - 1) == '\n');
             return true;
         }
 
         /// <summary>
-        /// Read signed 64 bit number
+        /// Tries to read a RESP length header from the given ASCII-encoded RESP string
+        /// and, if successful, moves the given ptr to the end of the length header.
         /// </summary>
-        public static bool Read64Int(out long number, ref byte* ptr, byte* end)
+        /// <param name="length">If parsing was successful, contains the extracted length from the header.</param>
+        /// <param name="ptr">The starting position in the RESP string. Will be advanced if parsing is successful.</param>
+        /// <param name="end">The current end of the RESP string.</param>
+        /// <param name="isArray">Whether to parse an array length header ('*...\r\n') or a string length header ('$...\r\n').</param>
+        /// <returns>True if a length header was successfully read.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool ReadLengthHeader(out int length, ref byte* ptr, byte* end, bool isArray = false)
         {
-            number = 0;
-            if (ptr + 3 >= end)
+            length = -1;
+            if (ptr + 3 > end)
                 return false;
 
-            Debug.Assert(*ptr == ':');
+            var readHead = ptr + 1;
+            var negative = *readHead == '-';
 
-            ptr++;
-            number = *ptr++ - '0';
-            while (*ptr != '\r')
+            // String length headers must start with a '$', array headers with '*'
+            if (*ptr != (isArray ? '*' : '$'))
             {
-                Debug.Assert(*ptr >= '0' && *ptr <= '9');
-                number = number * 10 + *ptr++ - '0';
-                if (ptr >= end)
+                RespParsingException.ThrowUnexpectedToken(*ptr);
+            }
+
+            // Special case: '$-1' (NULL value)
+            if (negative)
+            {
+                if (readHead + 4 > end)
+                {
                     return false;
+                }
+                if (*(uint*)readHead == MemoryMarshal.Read<uint>("-1\r\n"u8))
```

Comment:
Are we expecting a command to have a null length header?
Will the following input result in a crash?
*2\r\n$3\r\nGET\r\n$-1\r\nfoo\r\n.

Reply:
I assume you are asking if the NULL value should only be parsed for commands that opt in, right? I think that's a very good question. `$-1` should not be allowed for GET, but currently will not result in an error message. If we make NULL opt-in, we could return an `InvalidStringLength` exception. Any thoughts on this?